### ISSUE TITLE
Support for Avalonia 0.10

### ISSUE THREAD

**Issue by apkrymov** [swharden/ScottPlot#571]
System.MissingMethodException: "Method not found: 'Void Avalonia.Controls.Image.set_Source(Avalonia.Media.Imaging.IBitmap)'." occured when I try to run application on Avalonia 10.0 instead 9.11.

Are there any plans to make the library compatible with the latest version of Avalonia?

The problem in AvaPlotBackend.cs.
```c#
public override void SetImagePlot(bool lowQuality)
{
    view.Find<Ava.Controls.Image>("imagePlot").Source = BmpImageFromBmp(plt.GetBitmap(true, lowQuality));
}
```

**Comment by swharden** [comment]
@Benny121221 (#496) I haven't looked into this yet... do you have any suggestions?

**Comment by Benny121221** [comment]
Firstly, sorry for the slow response, midterms are getting wild.

@apkrymov The issue does not appear when the `ScottPlot.Avalonia` library is compiled with Avalonia 0.10.0-preview5, the most recent preview version available as long as you remove/comment out this [line](https://github.com/swharden/ScottPlot/blob/master/src/ScottPlot.Avalonia/HelpWindow.axaml.cs#L16). Since Avalonia 0.10 is in preview I don't think it makes sense to publish a new release to nuget for support unless @swharden wants to go down the road of shipping preview versions to nuget (which I suspect he doesn't but he may surprise me).

Hopefully by the time Avalonia reaches 1.0 there'll be better compatibility between versions but this is what 0.x versions are for, they're expected to change often. In case you're curious the issue was caused by the relevant function going from taking an `IBitmap` to an `IImage`.

So in summary, I've set up github to give me a notification on new Avalonia releases and when 0.10.0 comes out of preview I'll make a pull request to add support. It seems like it's only changing the dependencies and removing one line. If Scott is ok with shipping preview releases to NuGet it may happen sooner but that could mean a lot more work for him. If you need Avalonia 0.10.0 features right away then compiling the library your self probably makes the most sense, and I suppose at least we know that it does work with only the minor changes above. Sorry for any inconvenience.

**Comment by swharden** [comment]
Hi @Benny121221, no worries about the delay! Thank you for your thorough answer.

### Supperted Version

Re-stating @Benny121221's answer, ScottPlot currently targets/supports Avalonia 0.9.11 as defined by these lines. These lines can be changed to control which version of Avalonia is used when the `ScottPlot.Avalonia` NuGet package is installed.

https://github.com/swharden/ScottPlot/blob/600d3d9d0486a9fea34d241b04f629a98bfc78da/src/ScottPlot.Avalonia/ScottPlot.Avalonia.NUGET.csproj#L31-L32

I agree that we shouldn't publish NuGet packages that target preview versions of Avalonia, so this line can remain as it is. When new version of Avalonia are officially released, this line can be changed accordingly. Similarly, I didn't add .NET Core WinForms support while that was in preview (and I'm glad! developing to target buggy preview environments is super frustrating). So, in all, I like the plain of waiting 👍

### Dev Tools

@Benny121221 I'm not entirely sure about the significance of this section.

https://github.com/swharden/ScottPlot/blob/600d3d9d0486a9fea34d241b04f629a98bfc78da/src/ScottPlot.Avalonia/HelpWindow.axaml.cs#L15-L17

https://avaloniaui.net/docs/quickstart/devtools

What is the relationship between the developer tools and @apkrymov's problem? If this line causes problems for people who upgrade Avalonia versions, should it be commented-out now so it will be gone in the next NuGet package, even though it targets 0.9? I publish a new `ScottPlot.Avalonia` package every time I publish `ScottPlot` (about once a week) so this change could be made pretty soon.

**Comment by Benny121221** [comment]
> What is the relationship between the developer tools and apkrymov's problem?

The project will not compile on Avalonia 0.10 with that line enabled. Why? I don't know, I couldn't find any changes to the relevant files that would break it. Perhaps the extension doesn't support 0.10.0 yet? I don't see why that would block compiling but ¯\_(ツ)_/¯

Also, this line needs to be commented out/removed in `ScottPlot.Demo.Avalonia`:

https://github.com/swharden/ScottPlot/blob/master/src/ScottPlot.Demo.Avalonia/Program.cs#L4

It's an import for an unused dependency (which now doesn't exist).

**Comment by swharden** [comment]
Okay cool. It sounds like the plan is I will delete:

https://github.com/swharden/ScottPlot/blob/600d3d9d0486a9fea34d241b04f629a98bfc78da/src/ScottPlot.Avalonia/HelpWindow.axaml.cs#L15-L17

and I'll delete:

https://github.com/swharden/ScottPlot/blob/cdedc1d2bd9eee3511c1565c8f33ec3c75f51916/src/ScottPlot.Demo.Avalonia/Program.cs#L4

... then the Avalonia control will be a bit more future-proof even though we leave it to target `0.9.11`. Once this is published on NuGet @apkrymov's problem will be solved too.

**Comment by apkrymov** [comment]
Great solution! It really works. @Benny121221, @swharden thank you for your quick response 😀

**Comment by swharden** [comment]
... just realized removing serilog broke the build

```
src\ScottPlot.Demo.Avalonia\Program.cs(19,18): Error CS1061: 'AppBuilder' does not contain a definition for 'LogToDebug' and no accessible extension method 'LogToDebug' accepting a first argument of type 'AppBuilder' could be found (are you missing a using directive or an assembly reference?)
```

I'll try deleting this line...

https://github.com/swharden/ScottPlot/blob/cdedc1d2bd9eee3511c1565c8f33ec3c75f51916/src/ScottPlot.Demo.Avalonia/Program.cs#L20

**Comment by Benny121221** [comment]
I think there's a decent likelihood that the recommended templates for Avalonia will change for 0.10 given that the current ones don't work. It may be easier to just wait until the new ones are documented than try and guess what they might be. I'll confess that I'm not a contributor to Avalonia and I don't actively follow its development, otherwise I might know what they recommend for testing the previews.

I notice that the checks pass but developer tools and logging probably should be reenabled for a future ScottPlot release. So this is probably easier to change once Avalonia 0.10 comes out.

**Comment by swharden** [comment]
> I'm not a contributor to Avalonia and I don't actively follow its development, otherwise I might know what they recommend for testing the previews.

It seems to be quite the moving target! I have some shared frustration here, and supporting multiple versions is a lot of work...

> I notice that the checks pass but developer tools and logging probably should be reenabled for a future ScottPlot release. So this is probably easier to change once Avalonia 0.10 comes out.

What makes most sense in production code? Perhaps developers could be encouraged to run from source (where they can add logging and debugging in themselves), but the NuGet package could omit these things to reduce friction for people who just want to publish a library or app that uses the control?

**Comment by Benny121221** [comment]
> Perhaps developers could be encouraged to run from source...
> ...for people who just want to publish a library or app that uses the control?

To me these are the same people? Developers are the ones that publish libraries and apps? You could try to encourage people to build from source for development but I don't expect many people are going to do that; I don't expect that the majority of people who install it from NuGet are going to look for anything but documentation in this repo. No offence but I did the exact same when I first installed ScottPlot.

To me, common sense would dictate that if a code snippet was put forwards as a template for general use in all Avalonia projects then the code is suitable for production. If it weren't I don't know why it would be in the Visual Studio template for Avalonia as well as on their website. Or it would be enclosed in a preprocessor directive like:
```cs
#if DEBUG
//Code
#endif
```

This tells me that logging is disabled when compiled with `RELEASE` defined or when there is no suitable logging target. Which means it should never be an issue in production code. To me, it looks like it's the latter, as it goes through `System.Diagnostics.Debug` according to http://reference.avaloniaui.net/api/Avalonia.Logging.Serilog/SerilogExtensions/F8867CA4. Because of this I don't think there's much merit in removing logging and forcing people to compile from sauce to add it back in, as it has no effect on production code.

**Comment by swharden** [comment]
Hi @Benny121221, your message is very well received! Regarding Avalonia project templates and convention, I don't have much Avalonia experience and am happy to proceed based on your recommendation. It sounds like your recommendation is to put the debug lines (and Serilog dependency) back in the way they were.

### Release vs. Debug

> logging is disabled when compiled with RELEASE defined or when there is no suitable logging target

Packages I upload NuGet are release builds. I think it's only when users run from source that they have the option to engage the debug option. This is why mentally I regard people running from source as "developers" (with access to a DEBUG build) and people using the NuGet packages as "production" users (forced to use the RELEASE build), but I recognize the language is awkward because everyone using these libraries is a developer. Hopefully this clears things up a bit!

```bat
:: This is the command I use to upload the Avalonia package to NuGet
nuget push src\ScottPlot.Avalonia\bin\Release\*.nupkg
```

**Comment by FoggyFinder** [comment]
> The project will not compile on Avalonia 0.10 with that line enabled. Why? I don't know, I couldn't find any changes to the relevant files that would break it.

It was moved into the separate package:

[Avalonia.Diagnostics](https://www.nuget.org/packages/Avalonia.Diagnostics)

You can find list of breaking changes here:

https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes#09---010

**Comment by swharden** [comment]
> It was moved into the separate package:

Ah, that makes sense. Thank you @FoggyFinder!

@Benny121221 want to fix this in the way you think is best as part of #576? (I'll close this issue now).

Note that the Avalonia control has two `csproj` files (one for the dev environment and another for NuGet) so if you want to change the required packages in one you'll probably want to make the change in both.

https://github.com/swharden/ScottPlot/blob/330a19d2f0ee6d432abbcf1d869fe55a5047c086/src/ScottPlot.Avalonia/ScottPlot.Avalonia.csproj#L6-L10

https://github.com/swharden/ScottPlot/blob/330a19d2f0ee6d432abbcf1d869fe55a5047c086/src/ScottPlot.Avalonia/ScottPlot.Avalonia.NUGET.csproj#L30-L37

**Comment by Benny121221** [comment]
> @Benny121221 want to fix this in the way you think is best as part of #576? (I'll close this issue now).

Well, I have to wait anyways unless you want to use the preview build, I can't do anything now. So in that sense I'm already done.
